### PR TITLE
PHP: Separate unit and integration tests

### LINKED ISSUE #5473: PHP Tests: Separating unit and integration tests
## Task Description

Might be worth looking into splitting up our current PHP integration tests into proper unit tests (with all of WP mocked) and separate integration tests.

Listing pros and cons, etc.

Related links:
https://github.com/Yoast/wordpress-seo
https://github.com/Yoast/wp-test-utils
https://giuseppe-mazzapica.gitbook.io/brain-monkey/

Some possible pros:

* Easier to extract/port non-WordPress specific code in the future if needed (decoupling)
* Better test performance
* Easier maintenance

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

#8792 shows that there's benefit in splitting up our PHP tests into unit + integration tests to have a clearer separation, less buggy tests and better performance.

## Summary

This PR moves integration tests to `tests/phpunit/integration` and creates `tests/phpunit/unit` for the unit tests.

**Depends on #8792 being merged first**

## Relevant Technical Choices

Updates test bootstrap as documented by [Yoast/wp-test-utils](https://github.com/Yoast/wp-test-utils).

## To-do

- Add more unit tests, either by writing new ones or moving over some from the integration test suite where applicable.

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #5473

### CODE REVIEW

**Comment by spacedmonkey** [comment]
I think we can also move the following to unit test. 

- [tests/phpunit/tests/AMP](https://github.com/google/web-stories-wp/tree/main/tests/phpunit/tests/AMP)
- [tests/phpunit/tests/Infrastructure](https://github.com/google/web-stories-wp/tree/main/tests/phpunit/tests/Infrastructure)
- [tests/phpunit/tests/Web_Stories_Compatibility.php](https://github.com/google/web-stories-wp/tree/main/tests/phpunit/tests/Web_Stories_Compatibility.php)

**Comment by spacedmonkey** [comment]
Also this PR needs to rebased.

File: .github/workflows/tests-unit-php.yml [hunk 1]
```diff
@@ -144,19 +144,22 @@ jobs:
 
       - name: Run tests
         run: |
-          npm run test:php
```
Comment (spacedmonkey) [inline .github/workflows/tests-unit-php.yml:147]: This file should be renamed to include integration. So tests-unit-integration-php.yml. Also update title of the workflow.

File: package.json [hunk 2]
```diff
@@ -204,8 +204,10 @@
     "test:karma:dashboard": "karma start karma-dashboard.config.cjs --single-run",
     "test:karma:dashboard:retry-failed": "karma start karma-dashboard.config.cjs --single-run --retry-failed",
     "test:karma:dashboard:watch": "karma start karma-dashboard.config.cjs --debug",
-    "test:php": "vendor/bin/phpunit",
-    "test:php:multisite": "vendor/bin/phpunit -c phpunit-multisite.xml.dist",
+    "test:php": "npm-run-all --parallel test:php:*",
+    "test:php:unit": "vendor/bin/phpunit",
+    "test:php:integration": "vendor/bin/phpunit -c phpunit-integration.xml.dist",
+    "test:php:multisite": "vendor/bin/phpunit -c phpunit-integration-multisite.xml.dist",
```
Comment (spacedmonkey) [inline package.json]:
```suggestion
    "test:php:integration:multisite": "vendor/bin/phpunit -c phpunit-integration-multisite.xml.dist",
```

File: .github/workflows/tests-unit-php.yml (hunk 1 above)
Comment (swissspidy) [reply .github/workflows/tests-unit-php.yml:147]: Neither would be technically correct as the workflow right now runs both unit and integration tests. Are you perhaps suggesting creating two separate workflows, `tests-php-integration.yml` and `tests-php-unit.yml`?

**Comment by swissspidy** [comment]
> * [tests/phpunit/tests/AMP](https://github.com/google/web-stories-wp/tree/main/tests/phpunit/tests/AMP?rgh-link-date=2021-09-02T08%3A51%3A06Z)

The `Output_Buffer` class uses `is_singular()`, so that one couldn't be moved. (Edit: or I just stub the function)

But the various sanitizers could be moved indeed. The `Canonical_Sanitizer` class + test need a little refactor to make that happen (uses `get_permalink()` currently), but that's trivial.

> * [tests/phpunit/tests/Infrastructure](https://github.com/google/web-stories-wp/tree/main/tests/phpunit/tests/Infrastructure?rgh-link-date=2021-09-02T08%3A51%3A06Z)

`ServiceBasedPlugin` heavily relies on `add_filter`/`apply_filters` logic for filtering services, so that one can't really be moved. But the other tests can definitely be moved.

> * [tests/phpunit/tests/Web_Stories_Compatibility.php](https://github.com/google/web-stories-wp/tree/main/tests/phpunit/tests/Web_Stories_Compatibility.php?rgh-link-date=2021-09-02T08%3A51%3A06Z)

The class relies on `get_bloginfo()` and `Yoast/wp-test-utils` only stubs it for two cases:

https://github.com/Yoast/wp-test-utils/blob/c5cdabd58f2aa6f2a93f734cf48125f880c90101/src/BrainMonkey/YoastTestCase.php#L26-L36

So I'd first need to check whether I can override/extend that stub.

**Comment by spacedmonkey** [comment]
Could we use [wp_mock](https://github.com/10up/wp_mock) to mock these functions? 

We could exact this [line](https://github.com/google/web-stories-wp/blob/main/includes/compat/Web_Stories_Compatibility.php#L122) and inject this into the class as a contractor.

**Comment by swissspidy** [comment]
No need for wp_mock, this can be done with BrainMonkey

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Once this is fixed https://github.com/google/web-stories-wp/pull/8818/files#r700930307 and conflicts are fixed, I will approve.

File: .github/workflows/tests-unit-php.yml (hunk 1 above)
Comment (spacedmonkey) [reply .github/workflows/tests-unit-php.yml:147]: Adding as two different workflow feel a little wasteful, specially as you need to do a composer install and run in different php versions. 

But wonder, if we just rename the file and title.

**Review by spacedmonkey — APPROVED** [review]
👍

**Pull request merged**